src/passes/OptimizeCasts.cpp

```diff
@@ -102,6 +120,273 @@ namespace wasm {
+  // where we cannot move casts past.
+  void updateRefCastMoves(size_t index) {
+    if (earliestRefCastReachable[index]) {
+      auto& applyVector = refCastToApply[earliestRefCastReachable[index]];
+      for (size_t i = 0; i < movedRefCast[index].size(); i++) {
+        if (!movedRefCast[index][i].atTarget) {
+          applyVector.push_back(movedRefCast[index][i]);
+        }
+      }
+      movedRefCast[index].clear();
+      earliestRefCastReachable[index] = nullptr;
+    }
+  }
+
+  // Same as above but for RefAs.
+  void updateRefAsMoves(size_t index) {
+    if (earliestRefAsReachable[index]) {
+      auto& applyVector = refAsToApply[earliestRefAsReachable[index]];
+      for (size_t i = 0; i < movedRefAs[index].size(); i++) {
+        if (!movedRefAs[index][i].atTarget) {
+          applyVector.push_back(movedRefAs[index][i]);
+        }
+      }
+      movedRefAs[index].clear();
+      earliestRefAsReachable[index] = nullptr;
+    }
+  }
+
+  static void doNoteNonLinear(EarlyCastFinder* self, Expression**) {
+    for (size_t i = 0; i < self->numLocals; i++) {
+      self->updateRefCastMoves(i);
+      self->updateRefAsMoves(i);
+    }
+  }
+
+  void visitExpression(Expression* curr) {
+    // A new one is instantiated for each expression to determine
+    // if a cast can be moved past it.
+    ShallowEffectAnalyzer currAnalyzer(options, *getModule(), curr);
+
+    if (testRefCast.invalidates(currAnalyzer)) {
+      for (size_t i = 0; i < numLocals; i++) {
+        updateRefCastMoves(i);
+      }
+    }
+
+    if (testRefAs.invalidates(currAnalyzer)) {
+      for (size_t i = 0; i < numLocals; i++) {
+        updateRefAsMoves(i);
+      }
+    }
+  }
+
+  void visitLocalSet(LocalSet* curr) {
+    visitExpression(curr);
+    updateRefCastMoves(curr->index);
+    updateRefAsMoves(curr->index);
+  }
+
+  void visitLocalGet(LocalGet* curr) {
+    visitExpression(curr);
+
+    if (!earliestRefCastReachable[curr->index]) {
+      earliestRefCastReachable[curr->index] = curr;
+    }
+
+    if (!earliestRefAsReachable[curr->index]) {
+      earliestRefAsReachable[curr->index] = curr;
+    }
+  }
+
+  void visitRefAs(RefAs* curr) {
+    visitExpression(curr);
+
+    auto* fallthrough = Properties::getFallthrough(curr, options, *getModule());
+
+    if (auto* get = fallthrough->dynCast<LocalGet>()) {
+      if (get != earliestRefAsReachable[get->index]) {
+        bool alreadyAdded = false;
+        for (size_t i = 0; i < movedRefAs[get->index].size(); i++) {
+          if (movedRefAs[get->index][i].op == curr->op) {
+            alreadyAdded = true;
+            break;
+          }
+        }
+
+        if (!alreadyAdded) {
+          movedRefAs[get->index].emplace_back(
+            curr->op, get == earliestRefAsReachable[get->index]);
+        }
+      }
+    }
+  }
+
+  void visitRefCast(RefCast* curr) {
+    visitExpression(curr);
+    auto* fallthrough = Properties::getFallthrough(curr, options, *getModule());
```

Comment:
After offline discussion I think fallthrough here is ok, but please add a detailed comment summarizing what we talked about (i.e., why it is ok - why the various possible situations end up not as problems).

Reply:
Added.